### PR TITLE
Modernized module in PhysicsTools/HepMCCandAlgos

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

- converted to thread-friendly module types
- consolidated headers into .cc files
- added fillDescriptions to some modules
- minor code improvements

#### PR validation:

Code compiles. Compiling with `-DUSE_CMS_DEPRECATED` no longer generates any warnings.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
Please test

File: PhysicsTools/HepMCCandAlgos/test/HZZ4muAnalyzer.h [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
+#include "FWCore/Framework/interface/one/EDAnalyzer.h"
 
 #include "SimDataFormats/GeneratorProducts/interface/HepMCProduct.h"
 
 // forward declarations
 class TFile;
 class TH1D;
 
-class HZZ4muAnalyzer : public edm::EDAnalyzer {
+class HZZ4muAnalyzer : public edm::one::EDAnalyzer<> {
```
Comment (slava77) [inline PhysicsTools/HepMCCandAlgos/test/HZZ4muAnalyzer.h:12]: there is also GeneratorInterface/Pythia6Interface/test/HZZ4muAnalyzer.cc

while this one becomes a plugin via
```C++
typedef HZZ4muAnalyzer HZZ4muTestAnalyzer;
DEFINE_FWK_MODULE(HZZ4muTestAnalyzer);
```

wouldn't this still potentially lead to symbol conflicts; perhaps more certain if someone attempts to load both analyzers.

the GeneratorInterface one goes as 
```C++
typedef HZZ4muAnalyzer HZZ4muExampleAnalyzer;
DEFINE_FWK_MODULE(HZZ4muExampleAnalyzer);
```

Looking in history, the only use case of this analyzer in PhysicsTools/HepMCCandAlgos/test/GenH190ZZ4muHepMCAnalysis.py was removed in #10030

OTOH, the other one in GeneratorInterface did not even go through the `consumes` migration.

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/36359	967b8d9 
- code changes are technical
    - GenPlusSimParticleProducer.cc would benefit from a fillDescriptions, but I guess this could be done in a separate PR
    - switching to `edm::global` in CommonTools/UtilAlgos/interface/Matcher.h and CommonTools/UtilAlgos/interface/NewMatcher.h is a bit iffy considering some other (not the ones currently in the release) template class instances can drag in tools that are not thread safe. Staying with `edm::stream` may be more future proof, but it's likely OK as proposed here.
- jenkins tests pass and comparisons with the baseline show no differences

**Comment by slava77** [comment]
@qliphy @perrotta 
you may want to bypass the analysis signature here.

File: PhysicsTools/HepMCCandAlgos/plugins/ModelFilter.cc [hunk 2]
```diff
@@ -1,11 +1,46 @@
...
+    explicit ModelFilter(const edm::ParameterSet&);
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+    typedef std::vector<std::string>::const_iterator comments_const_iterator;
+
+  private:
+    bool filter(edm::StreamID, edm::Event&, const edm::EventSetup&) const override;
+    static std::vector<std::string> split(std::string const& fstring, std::string const& splitter);
+
+    edm::EDGetTokenT<LHEEventProduct> tokenSource_;
```
Comment (perrotta) [inline PhysicsTools/HepMCCandAlgos/plugins/ModelFilter.cc:36]: All these parameters could be made `const`, and the same applies to several other classes touched by this PR.
I imagine that all these kind of (non mandatory) optimizations can be defereed to another PR, if wanted.

**Comment by perrotta** [comment]
There are several optimizations (and fixes, see what reported by @slava77 in https://github.com/cms-sw/cmssw/pull/36359#discussion_r767043242) that could be applied to the code touched in this PR.

At the first order, this could be even accepted as such. Before doing so, @Dr15Jones could you please comment about what Slava writes in https://github.com/cms-sw/cmssw/pull/36359#issuecomment-991374823, i.e. "switching to edm::global in CommonTools/UtilAlgos/interface/Matcher.h and CommonTools/UtilAlgos/interface/NewMatcher.h is a bit iffy considering some other (not the ones currently in the release) template class instances can drag in tools that are not thread safe"?

**Comment by Dr15Jones** [comment]
> switching to edm::global in CommonTools/UtilAlgos/interface/Matcher.h and CommonTools/UtilAlgos/interface/NewMatcher.h is a bit iffy considering some other (not the ones currently in the release) template class instances can drag in tools that are not thread safe. Staying with edm::stream may be more future proof, but it's likely OK as proposed here.

The existing interface was already overriding a `const` virtual function so the only way it gets thread-unsafe would be to use mutable or a non-const global variable, both of which we tell people not to do. Therefore, existing non CMSSW code is unlikely to be affected.

**Comment by perrotta** [comment]
+1
- @cms-sw/analysis-l2 , while we are overriding your signature here to speed up, you can perhaps have a glance at the issue reported above in https://github.com/cms-sw/cmssw/pull/36359#discussion_r767043242 , with possible conflicts in plugin names: some cleaning of possibly unused files may also avoid unneeded future maintenance of the same

**Comment by perrotta** [comment]
merge

**Comment by santocch** [comment]
+1
